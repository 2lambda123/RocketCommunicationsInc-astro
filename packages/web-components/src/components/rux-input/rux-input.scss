@use '../../common/functional-components/FormFieldMessage/form-field-message.scss';

:host {
    /**
  * @prop --input-background-color: [DEPRECATED] the input background color
  */
    --input-background-color: var(--color-background-base-default);

    /**
    * @prop --input-border-color: [DEPRECATED] the input border color
    */
    --input-border-color: var(--color-border-interactive-muted);

    /**
    * @prop --input-text-color: [DEPRECATED] the input text color
    */
    --input-text-color: var(--color-text-primary);

    /**
    * @prop --input-focus-border-color: [DEPRECATED] the input focus border color
    */

    --input-focus-border-color: var(--color-background-interactive-hover);

    /**
    * @prop --input-selection-background-color: [DEPRECATED] the background color of highlighted text
    */
    --input-selection-background-color: var(--color-palette-brightblue-400);

    /**
    * @prop --input-invalid-border-color: [DEPRECATED] the input invalid border color
    */
    --input-invalid-border-color: var(--color-border-error);

    display: block;

    .rux-form-field {
        display: flex;
        flex-direction: column;
        align-items: flex-start;
        justify-content: center;
        color: var(--color-text-primary);
    }
    rux-icon {
        cursor: pointer;
    }

    input[type='date' i]::-webkit-calendar-picker-indicator {
        cursor: pointer;
        --webkit-appearance: none;
        background: var(--input-background-color)
            url("data:image/svg+xml,%3Csvg width='24' height='24' viewBox='0 0 24 24' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M19 4H18V3C18 2.45 17.55 2 17 2C16.45 2 16 2.45 16 3V4H8V3C8 2.45 7.55 2 7 2C6.45 2 6 2.45 6 3V4H5C3.89 4 3.01 4.9 3.01 6L3 20C3 21.1 3.89 22 5 22H19C20.1 22 21 21.1 21 20V6C21 4.9 20.1 4 19 4ZM19 19C19 19.55 18.55 20 18 20H6C5.45 20 5 19.55 5 19V9H19V19ZM9 13V11H7V13H9ZM11 11H13V13H11V11ZM17 13V11H15V13H17Z' fill='%234dacff'/%3E%3C/svg%3E%0A");
        background-position: center;
    }
    input[type='datetime-local' i]::-webkit-calendar-picker-indicator {
        cursor: pointer;
        --webkit-appearance: none;
        background: var(--input-background-color)
            url("data:image/svg+xml,%3Csvg width='24' height='24' viewBox='0 0 24 24' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M19 4H18V3C18 2.45 17.55 2 17 2C16.45 2 16 2.45 16 3V4H8V3C8 2.45 7.55 2 7 2C6.45 2 6 2.45 6 3V4H5C3.89 4 3.01 4.9 3.01 6L3 20C3 21.1 3.89 22 5 22H19C20.1 22 21 21.1 21 20V6C21 4.9 20.1 4 19 4ZM19 19C19 19.55 18.55 20 18 20H6C5.45 20 5 19.55 5 19V9H19V19ZM9 13V11H7V13H9ZM11 11H13V13H11V11ZM17 13V11H15V13H17Z' fill='%234dacff'/%3E%3C/svg%3E%0A");
        background-position: center;
    }
    input[type='time' i]::-webkit-calendar-picker-indicator {
        background: var(--input-background-color)
            url("data:image/svg+xml,%3Csvg width='24' height='24' viewBox='0 0 24 24' fill='none' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath fill-rule='evenodd' clip-rule='evenodd' d='M11.99 2C6.47 2 2 6.48 2 12C2 17.52 6.47 22 11.99 22C17.52 22 22 17.52 22 12C22 6.48 17.52 2 11.99 2ZM12 20C7.58 20 4 16.42 4 12C4 7.58 7.58 4 12 4C16.42 4 20 7.58 20 12C20 16.42 16.42 20 12 20ZM11.72 7H11.78C12.18 7 12.5 7.32 12.5 7.72V12.26L16.37 14.56C16.72 14.76 16.83 15.21 16.62 15.55C16.42 15.89 15.98 15.99 15.64 15.79L11.49 13.3C11.18 13.12 11 12.79 11 12.44V7.72C11 7.32 11.32 7 11.72 7Z' fill='%234dacff'/%3E%3C/svg%3E");
        cursor: pointer;
        background-position: center;
    }
    .rux-input {
        input {
            border: none;
            background: none;
            box-shadow: none;
            padding: 0;
            margin: 0;
            color: inherit;
            font-family: inherit;
            font-size: inherit;
            font-weight: inherit;
            -webkit-appearance: none;
            appearance: none;
            outline: none;
            width: 100%;
            &:disabled {
                cursor: not-allowed;
            }
        }
        width: 100%;
        overflow: hidden;
        flex: 1 1 auto;
        display: inline-flex;
        align-items: center;
        justify-content: start;
        position: relative;
        box-sizing: border-box;
        order: 2;
        border: 1px solid var(--input-border-color);
        border-radius: var(--radius-base);
        font-family: var(--font-body-1-font-family);
        font-size: var(--font-body-1-font-size);
        font-weight: var(--font-body-1-font-weight);
        letter-spacing: var(--font-body-1-letter-spacing);
        color: var(--input-text-color);
        background-color: var(--input-background-color);

        &--invalid {
            border: 1px solid var(--input-invalid-border-color);
        }
        &--disabled {
            opacity: 0.4;
            opacity: var(--disabled-opacity);
            cursor: not-allowed;
        }
        &--small {
            padding: 0.219rem 0.5rem; // 3.5px 8px;
            min-height: 1.625rem; //26px
        }
        &--medium {
            padding: 0.469rem 0.5rem; // 7.5px 8px;
            min-height: 2.25rem; //36px
        }
        &--large {
            padding: 0.844rem 0.5rem; //13.5px 8px;
            min-height: 2.875rem; //46px
        }
        &--search {
            -webkit-appearance: none;
            -moz-appearance: none;
            padding-left: 2.25rem; //36px
            background: var(--input-background-color)
                url("data:image/svg+xml,%3Csvg width='40' height='40' xmlns='http://www.w3.org/2000/svg'%3E%3Cpath d='M35.37 26.457a15.282 15.282 0 000-21.918c-6.176-6.052-16.187-6.052-22.361 0a15.274 15.274 0 00-1.541 20.166c-.367.147-.713.37-1.014.665L.926 34.709a3.056 3.056 0 000 4.383 3.208 3.208 0 004.472 0l9.528-9.339c.352-.345.604-.753.756-1.186 6.137 3.831 14.347 3.124 19.687-2.11zM24.193 4.043c6.454 0 11.686 5.129 11.686 11.455 0 6.326-5.232 11.455-11.686 11.455-6.455 0-11.687-5.129-11.687-11.455 0-6.326 5.232-11.455 11.687-11.455z' fill='%234dacff' fill-rule='evenodd'/%3E%3C/svg%3E")
                10px/0.975rem no-repeat;
        }
        &--focused {
            border-color: var(--input-focus-border-color);
            outline: none;
        }
        &:hover {
            border-color: var(--input-focus-border-color);
            outline: none;
        }
        &--disabled:hover {
            border: 1px solid var(--input-border-color);
            color: var(--input-text-color);
        }
        &--invalid:hover {
            border: 1px solid var(--input-invalid-border-color);
        }
        &::placeholder {
            color: var(--color-text-placeholder);
        }
        .rux-input-prefix,
        .rux-input-suffix {
            display: inline-flex;
            flex: 0 0 auto;
            align-items: center;
            cursor: default;
        }
        .rux-input-prefix {
            ::slotted(*) {
                padding-right: 10px;
            }
        }
        .rux-input-suffix {
            ::slotted(*) {
                padding-left: 10px;
            }
        }
    }

    .rux-input-label {
        font-family: var(--font-body-1-font-family);
        font-size: var(--font-body-1-font-size);
        font-weight: var(--font-body-1-font-weight);
        letter-spacing: var(--font-body-1-letter-spacing);
        margin-bottom: var(--spacing-input-label-top);
        &__asterisk {
            margin-left: 4px;
        }
    }

    ::selection {
        background-color: var(--input-selection-background-color);
    }
}
.pw-button {
    padding-top: 1px; //icon looked like half a px off from being centered.
}
.pw-button::part(container) {
    display: flex;
    flex-direction: column;
    align-items: flex-end;
    justify-content: center;
    height: 1rem;
}

.hidden,
:host([hidden]) {
    display: none;
}
