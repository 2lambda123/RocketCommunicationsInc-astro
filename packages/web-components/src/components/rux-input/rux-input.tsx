import {
    Prop,
    Host,
    Component,
    Event,
    EventEmitter,
    h,
    Element,
    State,
    Watch,
} from '@stencil/core'
import FormFieldMessage from '../../common/functional-components/FormFieldMessage/FormFieldMessage'
import { FormFieldInterface } from '../../common/interfaces.module'
import { hasSlot, renderHiddenInput } from '../../utils/utils'

let id = 0

/**
 * @slot label - The input label
 * @part error-text - The error text element
 * @part form-field - The form-field wrapper container
 * @part help-text - The help text element
 * @part icon - The icon displayed when toggle-password prop is set
 * @part input - The input element
 * @part label - The input label when `label` prop is set
 * @part required - The asterisk when required is true
 *
 */
@Component({
    tag: 'rux-input',
    styleUrl: 'rux-input.scss',
    shadow: true,
})
export class RuxInput implements FormFieldInterface {
    private inputId = `rux-input-${++id}`

    @Element() el!: HTMLRuxInputElement

    @State() hasLabelSlot = false

    @State() togglePassword = false

    @State() isPasswordVisible = false

    @State() iconName = 'visibility'

    /**
     * The input label text. For HTML content, use the `label` slot instead.
     */
    @Prop() label?: string
    /**
     * The input placeholder text
     */
    @Prop() placeholder?: string

    /**
     * The help or explanation text
     */
    @Prop({ attribute: 'help-text' }) helpText?: string

    /**
     * The validation error text
     */
    @Prop({ attribute: 'error-text' }) errorText?: string

    /**
     * Presentational only. Renders the Input Field as invalid.
     */
    @Prop() invalid = false

    /**
     * The input value
     */
    @Prop({ mutable: true, reflect: true }) value: string = ''

    /**
     * The input name
     */
    @Prop() name = ''

    /**
     * The input type
     */
    @Prop() type:
        | 'text'
        | 'number'
        | 'email'
        | 'url'
        | 'search'
        | 'password'
        | 'tel' = 'text'

    /**
     * The input min attribute
     */
    @Prop() min?: string

    /**
     * The input max attribute
     */
    @Prop() max?: string

    /**
     * Disables the button via HTML disabled attribute. Button takes on a distinct visual state. Cursor uses the not-allowed system replacement and all keyboard and mouse events are ignored.
     */
    @Prop({ reflect: true }) disabled = false

    /**
     * Sets the input as required
     */
    @Prop() required: boolean = false

    /**
     * Control the padding around the input field
     */
    @Prop() size: 'small' | 'medium' | 'large' = 'medium'

    /**
     * The input step attribute
     */
    @Prop() step?: string

    /**
     * The input's autocomplete attribute
     */
    @Prop() autocomplete?: string

    /**
     * The input's spellcheck attribute
     */
    @Prop() spellcheck = false

    /**
     * The inputs readonly attribute
     */
    @Prop() readonly = false

    /**
     * Fired when the value of the input changes - [HTMLElement/input_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/input_event)
     */
    @Event({ eventName: 'ruxchange' }) ruxChange!: EventEmitter

    /**
     * Fired when an alteration to the input's value is committed by the user - [HTMLElement/change_event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/change_event)
     */
    @Event({ eventName: 'ruxinput' }) ruxInput!: EventEmitter

    /**
     * Fired when an element has lost focus - [HTMLElement/blur_event](https://developer.mozilla.org/en-US/docs/Web/API/Element/blur_event)
     */
    @Event({ eventName: 'ruxblur' }) ruxBlur!: EventEmitter

    @Watch('label')
    handleLabelChange() {
        this._handleSlotChange()
    }

    @Watch('type')
    handleTypeChange() {
        this._setTogglePassword()
    }

    connectedCallback() {
        this._onChange = this._onChange.bind(this)
        this._onInput = this._onInput.bind(this)
        this._handleSlotChange = this._handleSlotChange.bind(this)
        this._handleType = this._handleType.bind(this)
        this._handleTogglePassword = this._handleTogglePassword.bind(this)
    }

    disconnectedCallback() {
        this.el!.shadowRoot!.removeEventListener(
            'slotchange',
            this._handleSlotChange
        )
    }

    componentWillLoad() {
        this._handleSlotChange()
        this._setTogglePassword()
    }

    get hasLabel() {
        return this.label ? true : this.hasLabelSlot
    }

    private _onChange(e: Event) {
        const target = e.target as HTMLInputElement
        this.value = target.value
        this.ruxChange.emit()
    }

    private _onInput(e: Event) {
        const target = e.target as HTMLInputElement
        this.value = target.value
        this.ruxInput.emit()
    }

    private _onBlur = () => {
        this.ruxBlur.emit()
    }

    private _handleSlotChange() {
        this.hasLabelSlot = hasSlot(this.el, 'label')
    }

    private _setTogglePassword() {
        if (this.type === 'password') {
            this.togglePassword = true
        }
    }

    private _handleTogglePassword() {
        this.isPasswordVisible = !this.isPasswordVisible
        if (this.isPasswordVisible) {
            this.iconName = 'visibility-off'
        } else {
            this.iconName = 'visibility'
        }
    }

    private _handleType() {
        let realType = ''
        !this.togglePassword
            ? (realType = this.type)
            : this.togglePassword && this.isPasswordVisible
            ? (realType = 'text')
            : (realType = this.type)
        return realType
    }

    render() {
        const {
            disabled,
            el,
            errorText,
            helpText,
            inputId,
            invalid,
            label,
            max,
            min,
            name,
            _onChange,
            _onInput,
            _onBlur,
            _handleType,
            _handleSlotChange,
            _handleTogglePassword,
            placeholder,
            required,
            step,
            type,
            value,
            hasLabel,
            iconName,
            size,
            autocomplete,
            spellcheck,
            readonly,
            togglePassword,
        } = this

        renderHiddenInput(true, el, name, value, disabled)
        return (
            <Host>
                <div class="rux-form-field" part="form-field">
                    <label
                        class={{
                            'rux-input-label': true,
                        }}
                        part="label"
                        aria-hidden={hasLabel ? 'false' : 'true'}
                        htmlFor={inputId}
                    >
                        <span
                            class={{
                                hidden: !hasLabel,
                            }}
                        >
                            <slot name="label" onSlotchange={_handleSlotChange}>
                                {label}
                                {required && (
                                    <span
                                        part="required"
                                        class="rux-input-label__asterisk"
                                    >
                                        &#42;
                                    </span>
                                )}
                            </slot>
                        </span>
                    </label>
                    <input
                        name={name}
                        disabled={disabled}
                        type={_handleType()}
                        aria-invalid={invalid ? 'true' : 'false'}
                        placeholder={placeholder}
                        required={required}
                        step={step}
                        min={min}
                        max={max}
                        value={value}
                        class={{
                            'rux-input': true,
                            'rux-input--disabled': disabled,
                            'rux-input--invalid': invalid,
                            'rux-input--search': type === 'search',
                            'rux-input--small': size === 'small',
                            'rux-input--medium': size === 'medium',
                            'rux-input--large': size === 'large',
                        }}
                        id={this.inputId}
                        spellcheck={spellcheck}
                        autocomplete={togglePassword ? 'off' : autocomplete}
                        readonly={readonly}
                        onChange={_onChange}
                        onInput={_onInput}
                        onBlur={_onBlur}
                        part="input"
                    ></input>
                    {togglePassword && (
                        <div
                            class={{
                                'icon-container': true,
                                'show-password': true,
                                'with-label': hasLabel,
                            }}
                        >
                            <rux-icon
                                exportparts="icon"
                                onClick={_handleTogglePassword}
                                icon={iconName}
                                size="extra-small"
                            />
                        </div>
                    )}
                </div>
                <FormFieldMessage
                    errorText={errorText}
                    helpText={helpText}
                ></FormFieldMessage>
            </Host>
        )
    }
}
