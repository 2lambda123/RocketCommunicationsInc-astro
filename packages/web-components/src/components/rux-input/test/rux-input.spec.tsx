import { newSpecPage } from '@stencil/core/testing'
import { RuxInput } from '../rux-input'

describe('rux-input', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [RuxInput],
            html: `<rux-input></rux-input>`,
        })
        expect(page.root).toEqualHtml(`
          <rux-input value="">
            <mock:shadow-root>
            <div class="rux-form-field" part="form-field">
              <label aria-hidden="true" class="rux-input-label" htmlfor="rux-input-1" part="label">
                <span class="hidden">
                  <slot name="label"></slot>
                </span>
              </label>
              <input aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-1" type="text" value="">
            </div>
            </mock:shadow-root>
            <input class="aux-input" type="hidden" value="">
          </rux-input>
      `)
    })

    it('renders label prop', async () => {
        const page = await newSpecPage({
            components: [RuxInput],
            html: `<rux-input label="hello"></rux-input>`,
        })
        expect(page.root).toEqualHtml(`
        <rux-input label="hello" value="">
          <mock:shadow-root>
          <div class="rux-form-field" part="form-field">
            <label aria-hidden="false" class="rux-input-label" htmlfor="rux-input-2" part="label">
              <span>
                <slot name="label">hello</slot>
              </span>
            </label>
            <input aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-2" type="text" value="">
          </div>
          </mock:shadow-root>
          <input class="aux-input" type="hidden" value="">
        </rux-input>
      `)
    })

    it('renders label slot', async () => {
        const page = await newSpecPage({
            components: [RuxInput],
            html: `<rux-input><div slot="label">hello</div></rux-input>`,
        })
        expect(page.root).toEqualHtml(`
        <rux-input value="">
          <mock:shadow-root>
          <div class="rux-form-field" part="form-field">
            <label aria-hidden="false" class="rux-input-label" htmlfor="rux-input-3" part="label">
              <span>
                <slot name="label"></slot>
              </span>
            </label>
            <input aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-3" type="text" value="">
          </div>
          </mock:shadow-root>
          <div slot="label">hello</div>
          <input class="aux-input" type="hidden" value="">
        </rux-input>
      `)
    })
    it('renders rux-icon if type is password', async () => {
        const page = await newSpecPage({
            components: [RuxInput],
            html: `<rux-input type='password'></rux-input>`,
        })
        expect(page.root).toEqualHtml(`
      <rux-input type="password" value>
      <mock:shadow-root>
        <div class="rux-form-field" part="form-field">
        <label aria-hidden="true" class="rux-input-label" htmlfor="rux-input-4" part="label">
              <span class="hidden">
                <slot name="label"></slot>
                </span>
              </label>
          <input value="" type="password" autocomplete="off" aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-4">
          </input>
          <div class="icon-container show-password">
            <rux-icon part="icon" icon="visibility" size="extra-small">
            </rux-icon>
          </div>
        </div>
      </mock:shadow-root>
      <input class="aux-input" type="hidden" value="">
      </rux-icon>
      `)
    })
})
