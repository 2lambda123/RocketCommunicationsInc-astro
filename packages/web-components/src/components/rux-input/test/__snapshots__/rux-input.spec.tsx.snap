// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rux-input renders 1`] = `
<rux-input value="">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <label aria-hidden="true" class="rux-input-label" htmlfor="rux-input-1" part="label">
        <span class="hidden">
          <slot name="label"></slot>
        </span>
      </label>
      <input aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-1" type="text" value="">
    </div>
  </mock:shadow-root>
  <input class="aux-input" type="hidden" value="">
</rux-input>
`;

exports[`rux-input renders label prop 1`] = `
<rux-input label="hello" value="">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <label aria-hidden="false" class="rux-input-label" htmlfor="rux-input-2" part="label">
        <span>
          <slot name="label">
            hello
          </slot>
        </span>
      </label>
      <input aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-2" type="text" value="">
    </div>
  </mock:shadow-root>
  <input class="aux-input" type="hidden" value="">
</rux-input>
`;

exports[`rux-input renders label slot 1`] = `
<rux-input value="">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <label aria-hidden="false" class="rux-input-label" htmlfor="rux-input-3" part="label">
        <span>
          <slot name="label"></slot>
        </span>
      </label>
      <input aria-invalid="false" class="rux-input rux-input--medium" id="rux-input-3" type="text" value="">
    </div>
  </mock:shadow-root>
  <div slot="label">
    hello
  </div>
  <input class="aux-input" type="hidden" value="">
</rux-input>
`;

exports[`rux-input renders rux-icon if type is password 1`] = `
<rux-input type="password" value="">
  <mock:shadow-root>
    <div class="rux-form-field" part="form-field">
      <label aria-hidden="true" class="rux-input-label" htmlfor="rux-input-4" part="label">
        <span class="hidden">
          <slot name="label"></slot>
        </span>
      </label>
      <input aria-invalid="false" autocomplete="off" class="rux-input rux-input--medium" id="rux-input-4" type="password" value="">
      <div class="icon-container show-password">
        <rux-icon icon="visibility" part="icon" size="extra-small"></rux-icon>
      </div>
    </div>
  </mock:shadow-root>
  <input class="aux-input" type="hidden" value="">
</rux-input>
`;
